schema/encoding/jsonschema/schema.go

```diff
@@ -18,132 +13,104 @@ var (
-		ew.writeFormat(`, "maxProperties": %s`, strconv.FormatInt(int64(s.MaxLen), 10))
+		m["maxProperties"] = s.MaxLen
 	}
-
-	if len(required) > 0 {
-		ew.writeFormat(`, "required": [%s]`, strings.Join(required, ", "))
+	if len(s.Fields) > 0 {
+		err = formatFields(m, s.Fields)
 	}
-	return ew.err
-}
 
-type fieldWriter struct {
-	errWriter
-	propertiesCount int
+	return err
 }
 
-// comma optionally outputs a comma.  Invoke this when you're about to write a property.  Tracks how many have been
-// written and emits if not the first.
-func (fw *fieldWriter) comma() {
-	if fw.propertiesCount > 0 {
-		fw.writeString(",")
-	}
-	fw.propertiesCount++
-}
+func formatFields(m map[string]interface{}, fields schema.Fields) error {
+	props := make(map[string]interface{}, len(fields))
+	required := make([]string, 0, len(fields))
 
-func (fw *fieldWriter) resetPropertiesCount() {
-	fw.propertiesCount = 0
-}
+	for key, field := range fields {
+		if field.Required {
+			required = append(required, key)
+		}
+		formatter, err := validatorFormatter(field.Validator)
+		if err != nil {
+			return err
+		}
+		fm, err := formatter.JSONSchema()
+		if err != nil {
+			return err
+		}
+		formatField(fm, field)
+		props[key] = fm
+	}
+	m["properties"] = props
 
-// sortedFieldNames returns a list with all field names alphabetically sorted.
-func sortedFieldNames(v schema.Fields) []string {
-	keys := make([]string, 0, len(v))
-	for k := range v {
-		keys = append(keys, k)
+	if len(required) > 0 {
+		sort.Strings(required)
+		m["required"] = required
 	}
-	sort.Strings(keys)
-	return keys
+	return nil
 }
 
-func encodeField(ew errWriter, key string, field schema.Field) error {
-	fw := fieldWriter{ew, 0}
-	fw.writeFormat("%q: {", key)
+func formatField(m map[string]interface{}, field schema.Field) {
 	if field.Description != "" {
-		fw.comma()
-		fw.writeFormat(`"description": %q`, field.Description)
+		m["description"] = field.Description
 	}
 	if field.ReadOnly {
-		fw.comma()
-		fw.writeFormat(`"readOnly": %t`, field.ReadOnly)
-	}
-	if field.Validator != nil {
-		// FIXME: This breaks if there are any Validators that may write nothing. E.g. a schema.Object with
-		// Schema set to nil. A better solution should be found before adding support for custom validators.
-		fw.comma()
-		fw.err = encodeValidator(ew, field.Validator)
+		m["readOnly"] = field.ReadOnly
 	}
 	if field.Default != nil {
-		b, err := json.Marshal(field.Default)
-		if err != nil {
-			return err
-		}
-		fw.comma()
-		fw.writeString(`"default": `)
-		fw.writeBytes(b)
+		m["default"] = field.Default
 	}
-	fw.writeString("}")
-	fw.resetPropertiesCount()
-	return fw.err
 }
 
-// encodeValidator writes JSON Schema keys and values based on v, without the outer curly braces, to w. Note
-// that not all FieldValidator types are supported at the moment.
-func encodeValidator(w io.Writer, v schema.FieldValidator) (err error) {
+func validatorFormatter(v schema.FieldValidator) (f Formatter, err error) {
```

Comment:
early return would probably be slightly clearer. I'm wary of relying on initialized and named returned values for this use case as it may trip up people later.